### DISCUSSION TITLE
Why I cannot set engine to "gpt-4" during my usage

### DISCUSSION THREAD

**Discussion by xuantianfengwu** [acheong08/ChatGPT#1437] [Q&A]
I found the engines in the file of "V3.py" has "gpt-4", thus I try to set the engine as "gpt-4" during my usage.

But then, I will meet 404 web request problem:
APIConnectionError: 404 Not Found {
    "error": {
        "message": "The model: `gpt-4` does not exist",
        "type": "invalid_request_error",
        "param": null,
        "code": "model_not_found"
    }
}

![MyIssue_gpt4](https://github.com/acheong08/ChatGPT/assets/26734562/2eabd97c-a885-4c09-a299-04d99e910d56)

Is there any explanation for the result? Why OpenAI returns a 404 for "gpt-4"?

**Comment by Gteditor99** [comment]
`print("Supported engines: ", ", ".join(tiktoken.model.MODEL_TO_ENCODING))` returns

Supported engines:  `gpt-4,` `gpt-3.5-turbo`, `text-davinci-003`, `text-davinci-002`, `text-davinci-001`, `text-curie-001`, `text-babbage-001`, `text-ada-001`, `davinci`, `curie`, `babbage`, `ada`, `code-davinci-002`, `code-davinci-001`, `code-cushman-002`, `code-cushman-001`, `davinci-codex`, `cushman-codex`, `text-davinci-edit-001`, `code-davinci-edit-001`, `text-embedding-ada-002`, `text-similarity-davinci-001`, `text-similarity-curie-001`, `text-similarity-babbage-001`, `text-similarity-ada-001`, `text-search-davinci-doc-001`, `text-search-curie-doc-001`, `text-search-babbage-doc-001`, `text-search-ada-doc-001`, `code-search-babbage-code-001`, `code-search-ada-code-001`, `gpt2`. 

It seems that any engine input (regardless of implementation) by the user returns a `NotImplemented` error?
```py
    def get_token_count(self, convo_id: str = "default") -> int:
        """
        Get token count
        """
        if self.engine not in ENGINES:
            raise NotImplementedError(f"Unsupported engine {self.engine}")

        tiktoken.model.MODEL_TO_ENCODING["gpt-4"] = "cl100k_base"

        encoding = tiktoken.encoding_for_model(self.engine)

        num_tokens = 0
        for message in self.conversation[convo_id]:
            # every message follows <im_start>{role/name}\n{content}<im_end>\n
            num_tokens += 5
            for key, value in message.items():
                num_tokens += len(encoding.encode(value))
                if key == "name":  # if there's a name, the role is omitted
                    num_tokens += 5  # role is always required and always 1 token
        num_tokens += 5  # every reply is primed with <im_start>assistant
        return num_tokens
```

**Comment by acheong08** [comment]
Your account doesn't have access to gpt-4. You need to apply on a waitlist
